services/image-rec-master/service/app.py

```diff
@@ -0,0 +1,213 @@
+
+
+@routes.get(r'/api/pipeline/targets/{id:\d+}')
+async def handle_get_pipeline_target_by_id(request):
+    """Return a target by id in the pipeline."""
+    target_id = int(request.match_info['id'])
+    target = await _get_target(request, target_id)
+
+    if target:
+        return _proto_response(request, target)
+    else:
+        return web.Response(status=404)
+
+
+@routes.post('/api/pipeline/reset')
+async def handle_reset_pipeline(request):
+    """Empty out the current Redis database to reset the pipeline."""
+    await request.app['redis'].flushdb()
+    return web.Response(status=204)
+
+
+@routes.get('/api/pipeline/archive')
+async def handle_get_pipeline_archive(request):
+    """Return an archive for backup target submission."""
+    odlcs = []
+    target_ids = await get_int_set(request.app['redis'], 'all-targets')
+
+    for target_id in target_ids:
+        target = await _get_target(request, target_id)
+
+        # Add the target if it hasn't errored or been removed.
+        if not target.errored and not target.removed:
+            odlcs.append(target.odlc)
+
+    archive = create_archive(odlcs)
+
+    if archive:
+        return web.Response(body=archive, headers={
+            'Content-Disposition': 'attachment; filename="targets.zip"',
+            'Content-Type': 'application/zip'
+        })
+    else:
+        return web.Response(status=204)
+
+
+async def _get_target(request, target_id):
+    target_hash = await request.app['redis'].hgetall(f'target:{target_id}')
+
+    if target_hash:
+        msg = Target()
+        msg.time = time()
+
+        msg.id = int(target_hash.get(b'id', b'0'))
+        msg.image_id = int(target_hash.get(b'image_id', b'0'))
+        msg.odlc.ParseFromString(target_hash.get(b'odlc', b''))
+        msg.submitted = target_hash.get(b'submitted', b'0') == b'1'
+        msg.errored = target_hash.get(b'errored', b'0') == b'1'
+        msg.removed = target_hash.get(b'removed', b'0') == b'1'
+
+        return msg
+    else:
+        return None
+
+
+def _proto_response(request, msg):
+    """Return a protobuf wire or JSON response."""
+    if not request.headers.getone('accept', '').startswith('application/json'):
+        body = msg.SerializeToString()
+        return web.Response(body=body, content_type='application/x-protobuf')
+    else:
+        body = MessageToJson(msg)
+        return web.json_response(body=body)
+
+
+def create_app():
+    """Create an aiohttp web application."""
+    app = web.Application()
+    app.on_startup.append(_start_tasks)
+    app.on_shutdown.append(_stop_tasks)
+    app.router.add_routes(routes)
+
+    return app
+
+
+async def _start_tasks(app):
+    app_tasks = []
+    app_tasks.append(_schedule_task(app, tasks.queue_new_images, 0.5))
+    app_tasks.append(_schedule_task(app, tasks.requeue_auto_images, 15.0))
+    app_tasks.append(_schedule_task(app, tasks.submit_targets, 0.0))
+    app_tasks.append(_schedule_task(app, tasks.remove_targets, 0.0))
```

Comment:
I think these parameters should be environment variables.

Reply:
Added a story for doing that in the future if needed.